### DISCUSSION TITLE
Communication with another IOC

### DISCUSSION THREAD

**Discussion by jdmax** [dls-controls/pythonSoftIOC#124] [Q&A] [answered]
I'm working on an application that will use several softiocs, and I'm try to figure out how to access the PVs from other IOCs. 

I would typically just "import epics" and use "epics.caget('PV_name')". But I've found that in a script that imports from softioc, I get an error when trying to do this: "cannot connect to 'PV_name'." For instance, the following script will work, unless I uncomment the softioc imports. Any insight on how to get or put to other IOCs would be very much appreciated.
```

import epics
import asyncio
#from softioc import softioc, builder, asyncio_dispatcher

async def main():

    print(epics.caget('Test_PV'))

if __name__ == "__main__":
    asyncio.run(main())
```

**Comment by AlexanderWells-diamond** [comment, marked as answer]
I unfortunately don't know exactly what package you are referring to when you say `import epics` - there appears to be no package on PyPi of that exact name. I can say that the likely issue is regarding threading. The default softioc will start a `cothread` callback thread, and the main process must occasionally yield to allow cothread to run.

The way to solve this is to use either [cothread](https://dls-controls.github.io/pythonSoftIOC/master/tutorials/creating-an-ioc.html) or [aioca](https://dls-controls.github.io/pythonSoftIOC/master/how-to/use-asyncio-in-an-ioc.html) depending on your preference for the `async` keyword - those documentation links should explain the differences.

**Reply by jdmax** [reply]
Thanks so much for your quick reply! The epics package I'm using is pyepics. I tried using cothread in the following example:

`import epics
import cothread
#from softioc import softioc, builder
import subprocess

def main():
    cothread.Spawn(print(epics.caget('Test_PV')))


if __name__ == "__main__":
    main()`

Again, the print works immediately if I keep the softioc import commented, but if I uncomment it and run again, the program hangs as it waits for the epics get, then returns "cannot connect to Test_PV". Thanks again for your help.

**Reply by AlexanderWells-diamond** [reply]
I see the same issue when using your example and installing pyepics.

The issue is that `pyepics` and `softioc` have two different sources of EPICS code libraries. `pyepics` appears to use [old, precompiled shared objects](https://github.com/pyepics/pyepics/tree/master/epics/clibs/linux64) whereas `pythonSoftIOC` uses [epicscorelibs](https://github.com/mdavidsaver/epicscorelibs). Trying to mix these two different sources of EPICS is effectively a non-starter - EPICS was never designed for this, and it causes problems all over the place.

You will need to use either cothread or [aioca](https://dls-controls.github.io/aioca/master/index.html) for the PV access inside pythonSoftIOC. I'd lean towards `aioca` as `cothread` has issues in Python3.11.

**Reply by jdmax** [reply]
Using aioca works well. I replaced the "epics.caget" with "await caget" and it is working as expected.  Thanks again!
